tsdb/index/postings.go

```diff
@@ -862,6 +863,11 @@ func (it *ListPostings) Err() error {
 	return nil
 }
 
+// Len returns the number of postings in the list.
+func (it *ListPostings) Len() int {
+	return it.length
```

Comment:
Wouldn't it easier to just say `return len(it.list)` here instead of adding a new field to the struct?

Reply:
That was my original implementation but `ListPostings.Next()` would advance and shrink the list so the length will change. Then to get the length of the list we have to call `Len` before calling `Next`, which I don't know might be bad for users.

```
	if len(it.list) > 0 {
		it.cur = it.list[0]
		it.list = it.list[1:]
		return true
	}
```

Another way I can think of is to use `it.cur + len(it.list)` as length.

Reply:
If we think it is fine to return length that is changing then the semantic of the API is changed. Idk if it is ok or not

Reply:
I see. My apologies. I haven't realized that we are iterating through the slice in this tricky way. (I don't know this code very well, just trying to get some reviews done in lack of people that are more qualified here.)

With a bit of research, I have seen that this was already attempted to write differently (with [an index variable](https://github.com/prometheus/prometheus/pull/12906)), but it turned out to slow down the intersection, so it was [reverted](https://github.com/prometheus/prometheus/pull/12955).

As this code path is so heavily optimized, maybe we have to be careful with adding an extra field to the struct. It increases its memory footprint by 25%.

@colega @bboreham WDYT?

Reply:
> If we think it is fine to return length that is changing then the semantic of the API is changed. 

I would expect the `Len()` to return the amount of postings I can still consume from this `ListPostings` instance. Knowing how many postings there were before I started consuming it doesn't give much value IMO.

(In reality, I can't imagine a use case where this would matter)

Reply:
Got it. Soudns good. Let me make the change.

If `Len()` method means the remaining postings legnth then if we add `Len()` method to other posting implementation in the future then we need to be consistent to that.

Reply:
OK, so let's do what I naively suggested before, but let's also change the doc comment to say "Len returns the number of remaining postings in the list."

Reply:
Updated in the latest commit.